tools/halide_benchmark.h

```diff
@@ -20,19 +39,129 @@ namespace Tools {
+
+struct BenchmarkConfig {
+  // Attempt to use this much time (in seconds) for the meaningful samples
+  // taken; initial iterations will be done to find an iterations-per-sample
+  // count that puts the total runtime in this ballpark.
+  double min_time{0.5};
+
+  // Run at least this many iterations per sample.
+  uint64_t min_iters{1};
+
+  // Run at most this many iterations over all samples.
+  uint64_t max_iters{kBenchmarkMaxIterations};
+
+  // Terminate when the relative difference between the best runtime
+  // seen and the third-best runtime seen is no more than
+  // this. Controls accuracy. The closer to zero this gets the more
+  // reliable the answer, but the longer it may take to run.
+  double accuracy{0.03};
+};
+
+struct BenchmarkResult {
+  // Best elapsed wall-clock time per iteration (seconds).
+  double wall_time;
+
+  // Number of samples used for measurement.
+  // (There might be additional samples taken that are not used
+  // for measurement.)
+  uint64_t samples;
+
+  // Total number of iterations across all samples.
+  // (There might be additional iterations taken that are not used
+  // for measurement.)
+  uint64_t iterations;
+
+  operator double() const { return wall_time; }
+};
+
+BenchmarkResult benchmark(std::function<void()> op, const BenchmarkConfig& config = {}) {
+  BenchmarkResult result{0, 0, 0};
+
+  const double min_time = std::max(10 * 1e-6, config.min_time);
+  const uint64_t min_iters = std::min(std::max((uint64_t)1, config.min_iters),
+                                      kBenchmarkMaxIterations);
+  const uint64_t max_iters = std::min(
+      std::max(config.min_iters, config.max_iters), kBenchmarkMaxIterations);
+  const double accuracy = 1.0 + std::min(std::max(0.001, config.accuracy), 0.1);
+
+  // We will do (at least) kMinSamples samples; we will do additional
+  // samples until the best the kMinSamples'th results are within the
+  // accuracy tolerance (or we run out of iterations).
+  constexpr int kMinSamples = 3;
+  double times[kMinSamples + 1] = {0};
+
+  double total_time = 0;
+  uint64_t iters_per_sample = min_iters;
+  while (result.iterations < max_iters) {
+    result.samples = 0;
+    result.iterations = 0;
+    total_time = 0;
+    for (int i = 0; i < kMinSamples; i++) {
+      times[i] = benchmark(1, iters_per_sample, op);
+      result.samples++;
+      result.iterations += iters_per_sample;
+      total_time += times[i] * iters_per_sample;
+    }
+    std::sort(times, times + kMinSamples);
+    if (times[0] * iters_per_sample * kMinSamples >= min_time) {
+      break;
+    }
+    // Use an estimate based on initial times to converge faster.
+    double next_iters = std::max(min_time / std::max(times[0] * kMinSamples, 1e-9),
+                                 iters_per_sample * 2.0);
+    iters_per_sample = (uint64_t)(next_iters + 0.5);
+  }
+
+  // While we aren't accurate enough (and we have time and iterations remaining)
```

Comment:
I don't think this condition is quite right. I think it should keep taking samples until it has the required accuracy, even if that goes beyond min_time. I was imagining:

(times[0] * accuracy < times[kMinSamples - 1] || total_time < min_time) && result.iterations < max_iters && total_time < max_time

Reply:
Yeah, good point: except there is no max_time setting. (max_iters is the "hard stop" limit.) Will fix.